pkg/router/router_test.go

```diff
@@ -0,0 +1,248 @@
+		routeClient: routeClient,
+		namespace:   namespace,
+	}
+
+	// Other shared junk
+	routerSelection := &routercmd.RouterSelection{}
+	factory := routerSelection.NewFactory(routeClient, projectClient.ProjectV1().Projects(), client)
+	informer := factory.CreateRoutesSharedInformer()
+	routeLister := routelisters.NewRouteLister(informer.GetIndexer())
+	lease := writerlease.New(time.Minute, 3*time.Second)
+	go lease.Run(wait.NeverStop)
+	tracker := controller.NewSimpleContentionTracker(informer, namespace, 60*time.Second)
+	tracker.SetConflictMessage(fmt.Sprintf("The router detected another process is writing conflicting updates to route status with name %q. Please ensure that the configuration of all routers is consistent. Route status will not be updated as long as conflicts are detected.", namespace))
+	go tracker.Run(wait.NeverStop)
+
+	var plugin router.Plugin
+
+	workdir, err := ioutil.TempDir("", "router")
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	fmt.Printf("router working directory: %s\n", workdir)
+
+	// The template plugin which is wrapped
+	svcFetcher := templateplugin.NewListWatchServiceLookup(client.CoreV1(), 60*time.Second, namespace)
+	pluginCfg := templateplugin.TemplatePluginConfig{
+		WorkingDir:            workdir,
+		DefaultCertificateDir: workdir,
+		ReloadFn:              func(shutdown bool) error { return nil },
+		TemplatePath:          "../../images/router/haproxy/conf/haproxy-config.template",
+		ReloadInterval:        15 * time.Second,
+	}
+	plugin, err = templateplugin.NewTemplatePlugin(pluginCfg, svcFetcher)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// Wrap the template plugin with other stuff
+	statusPlugin := controller.NewStatusAdmitter(plugin, routeClient.RouteV1(), routeLister, "default", "example.com", lease, tracker)
+	plugin = statusPlugin
+	plugin = controller.NewUniqueHost(plugin, routerSelection.DisableNamespaceOwnershipCheck, statusPlugin)
+	plugin = controller.NewHostAdmitter(plugin, routerSelection.RouteAdmissionFunc(), false, false, statusPlugin)
+
+	// Start the controller
+	c := factory.Create(plugin, false)
+	c.Run()
+
+	os.Exit(m.Run())
+}
+
+func TestAdmissionEdgeCases(t *testing.T) {
+	start := time.Now()
+
+	tests := map[string][]expectation{
+		"deletion promotes inactive routes": {
+			mustCreate{name: "a", host: "example.com", path: "", time: start},
+			mustCreate{name: "b", host: "example.com", path: "/foo", time: start.Add(1 * time.Minute)},
+			mustCreate{name: "c", host: "example.com", path: "/foo", time: start.Add(2 * time.Minute)},
+			mustCreate{name: "d", host: "example.com", path: "/foo", time: start.Add(3 * time.Minute)},
+			mustCreate{name: "e", host: "example.com", path: "/bar", time: start.Add(4 * time.Minute)},
+
+			expectAdmitted{"a", "b", "e"},
+			expectRejected{"c", "d"},
+
+			mustDelete{"b"},
+
+			expectAdmitted{"a", "c", "e"},
+			expectRejected{"d"},
+
+			mustDelete{"c"},
+
+			expectAdmitted{"a", "d", "e"},
+			expectRejected{},
```

Comment:
Is this intended to assert that there are 0 rejected routes?  It looks like it actually asserts nothing because `(e expectRejected) Apply` will iterate 0 times and return nil.

Reply:
Ah, yeah, I forgot to actually implement that and so the empty assertion is merely cosmetic at the moment. I'll fix it. Thanks!

Reply:
Removed those empty assertions.